frame/uniques/src/lib.rs

```diff
@@ -92,7 +92,12 @@ pub mod pallet {
 		type Event: From<Event<Self, I>> + IsType<<Self as frame_system::Config>::Event>;
 
 		/// Identifier for the collection of item.
-		type CollectionId: Member + Parameter + MaxEncodedLen + Copy;
+		type CollectionId: Member
+			+ Parameter
+			+ MaxEncodedLen
+			+ Copy
+			+ Default
+			+ AtLeast32BitUnsigned;
```

Comment:
@jsidorenko @Szegoo This trait bound here is breaking XCMv3, as we're using a `MultiLocation` for the `CollectionId`, which isn't an unsigned integer (but contains variants that can accept unsigned ints).

Reply:
@KiChjang Hmm, I will look more into this, but I am not even sure if it would be possible to have this auto-increment feature since as you said the `MultiLocation` is used for `CollectionId`, and not all variants store a uint.

Reply:
Looks like it is `MultiLocation` ordered, we could let the caller increment it off-chain.

Reply:
Maybe we could make this feature optional for some chains that don't support XCM and don't use `MultiLocation` for the`CollectionId`.

Reply:
Sorry but no, you're viewing this as an XCM problem rather than a problem with the underlying assumptions of this PR.

It is _not_ correct to automatically assume that all `CollectionId`s are unsigned integers. There are types like the `MultiLocation` enum where it has a variant that does contain an unsigned integer, and there could also be a struct acting as a `CollectionId` that contains a field with an unsigned integer type.

It is then simply _wrong_ to put the `AtLeast32BitUnsigned` trait bound on this associated type.

Reply:
Ok, I have one more idea, we could probably replace the `AtLeast32BitUnsigned` trait with some sort of `Incrementable` trait so that we don't make an assumption about the type of the `CollectionId`. And if the type of `CollectionId` cannot implement this trait(like `MultiLocation`) it should be possible to turn this feature off in the configuration.

If this doesn't make sense, then we should probably revert this change.

Reply:
Right, and this trait would be placed in the uniques pallet, as it's a requirement for its `CollectionId` associated type.

Whatever idea that you come up with, we'll absolutely need to remove this `AtLeast32BitUnsigned` trait bound.

Reply:
Yes, `AtLeast32BitUnsigned` definitely needs to be removed. I will open a new PR that implements the new idea and removes the `AtLeast32BitUnsigned` trait if you agree on this.

Reply:
But if we make it possible to disable the feature, then we have the same problem as before.  
I think we should rather revert and re-think.

Reply:
I was thinking of the following(this may be a bit hacky), make the `Incrementable` trait optional by making a default implementation of the `increment` function. Actually, as of writing this, this made me realize that this would pollute the code because we would need to add checks to the `create` and `try_increment_id` functions to decide how to act if the auto-increment feature is used. This would probably be pretty messy. So yeah. we should re-think this.

Reply:
Why does the type that this pallet assume affect XCM?

Why can't this pallet assume u32 ids, and then convert it to an appropriate multilocation?

Reply:
Take a look at the CI of this PR: https://github.com/paritytech/polkadot/pull/4097, specifically https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/1742302.

You'll see that it threw an error because `MultiLocation` is being specified as the concrete type for `CollectionId`, and it highlights `AtLeast32BitUnsigned` as the culprit.

> Why can't this pallet assume u32 ids, and then convert it to an appropriate multilocation?

There is no such conversion functionality in this PR, it has only assumed that the `CollectionId` is an unsigned int, and doesn't contain any logic to convert it back to a type that the pallet user specifies.

Reply:
I believe that @shawntabrizi meant to have the conversion login defined inside polkadot, would that make sense? This type of logic cannot be written inside the pallet, it should probably be defined in the runtimes.

Reply:
There is in fact a conversion code within XCM right now and it looks like this:

```rust
// `EnsureOriginWithArg` impl for `CreateOrigin` which allows only XCM origins
// which are locations containing the class location.
pub struct ForeignCreators;
impl EnsureOriginWithArg<Origin, MultiLocation> for ForeignCreators {
	type Success = AccountId;

	fn try_origin(o: Origin, a: &MultiLocation) -> sp_std::result::Result<Self::Success, Origin> {
		let origin_location = pallet_xcm::EnsureXcm::<Everything>::try_origin(o.clone())?;
		if !a.starts_with(&origin_location) {
			return Err(o)
		}
		SovereignAccountOf::convert(origin_location).map_err(|_| o)
	}

	#[cfg(feature = "runtime-benchmarks")]
	fn successful_origin(a: &MultiLocation) -> Origin {
		pallet_xcm::Origin::Xcm(a.clone()).into()
	}
}
```

The `MultiLocation` here is in fact the `CollectionId`, and so if we change it to `u128` instead of `MultiLocation`, then the code won't be able to verify whether the origin location is the owner/creator of the NFT collection.

In short, the `MultiLocation` type is used as an access control mechanism to see whether an incoming origin is allowed to modify a specific class of unique assets. Changing it to `u128` would not allow for such a mechanism to be implemented.

Reply:
I am still not convinced here. For example, the Balances pallet uses u128 as the balance type, not a MultiAsset. The type of a specific pallet should not affect the more complex wrappers needed for XCM.

There should simply be some trait which converts from `u32` -> `MultiLocation` for the assets pallet written into polkadot / cumulus or whatever.

Reply:
The point here is that the `CollectionId` is used to identify whether the asset collection is foreign via the `MultiLocation` struct. If we change `CollectionId` to be a `u128`, how exactly do we identify asset collections that are created by a foreign chain?

Honestly, I don't think I am the best person to be defending this design here. If you feel really strongly about this, I suggest that you talk with Gav about why the `CollectionId` _must_ be a `MultiLocation`, as he is the person who wrote the code.

Reply:
Okay, I understand now. if we want assets pallet to blindly accept teleports from any chain for any token, then we should have all collectionid be a multilocation.

Okay this makes sense. What we need to do then is have any assets which are created FROM the pallet, follow this increment behavior, but not the underlying token registry itself.

Reply:
In the general case, I still feel like it doesn't make sense for a `CollectionId` to always be an unsigned integer. `MultiLocation` is one example, but there's no reason why the following can't be a `CollectionId` either:

```rust
struct TaggedId {
    tag: BoundedVec<u8, ConstU32<16>>,
    id: u16,
}
// or
struct IdWrapper(u16)
// or even
(u128, u16)
```

These types all contain an unsigned integer as its field/member, and may contain crucial information to identify an asset collection. We should then not impose limits on `CollectionId` and force users to always use unsigned integers as the only kind of types for `CollectionId`.

Reply:
Should we create a Github discussion on this, if not what is the final agreement here?

Reply:
Does it mean the Statemint/e could receive and create a collection via XCM with a non-integer id? I'm pretty sure that would break all the UIs/marketplaces..
But if we can't really restrict the `CollectionId` field, then we should probably create a new internal auto-incremental field. Since the main issue here is how to restrict the creation of the same collection again, that internal field could help a bit.

P. S. In the UniquesV2 branch, instead of adding `AtLeast32BitUnsigned` trait, I've added `CheckedAdd`. Not sure if that could help with the XCM incompatibility problem.

Reply:
I've created a new PR that adds an internal field to track the auto-incremental collection's id (called id `ref_id`). Let's continue the discussion here: https://github.com/paritytech/substrate/pull/12166

Reply:
@KiChjang do you know how we could make the id param optional? When building a dapp, I never saw one asking a user to enter the id manually. I assume the dapp would simply try to send tons of rpc calls in order to find out an empty id on the background, which is not optimal.

Reply:
I think creating a new trait to bound the `CollectionId` sounds reasonable. This is basically the idea that @Szegoo had before using the `Incrementable` trait, and the basic idea here is to let the `Collectionid` type tell you what the next ID is, instead of making it the responsibility of this pallet.

Reply:
@Szegoo could you continue your work on `Incrementable`?

Reply:
Sure, I will continue on my work on `Incrementable`.

Reply:
I will continue working on this in #12057

Reply:
Super! And pls point into `js/uniques-v2-main-branch` afterwards

Reply:
Just to make sure I understand this correctly, we should add an `Incrementable` trait bound to the `CollectionId` as I suggested earlier, so this means that we are not going to store this internal `ref_id` field, right?

Reply:
This is how I understood it as well. @KiChjang Pls correct us if we're wrong

Reply:
That's correct, the idea is to let runtime configuration figure out what the next ID is, instead of having it become a concern for the uniques pallet itself.